extensions/aws/processors/PutKinesisStream.cpp

```diff
@@ -0,0 +1,201 @@
+  static constexpr uint64_t SINGLE_RECORD_MAX_SIZE = 1_MB;
+  std::unordered_map<std::string, StreamBatch> stream_batches;
+  uint64_t ff_count_in_batches = 0;
+  while (ff_count_in_batches < batch_size_) {
+    std::shared_ptr<core::FlowFile> flow_file = session.get();
+    if (!flow_file) { break; }
+    const auto flow_file_size = flow_file->getSize();
+    if (flow_file_size > SINGLE_RECORD_MAX_SIZE) {
+      flow_file->setAttribute(AwsKinesisErrorMessage.name, fmt::format("record too big {}, max allowed {}", flow_file_size, SINGLE_RECORD_MAX_SIZE));
+      session.transfer(flow_file, Failure);
+      logger_->log_error("Failed to publish to kinesis record {} because the size was greater than {} bytes", flow_file->getUUID().to_string(), SINGLE_RECORD_MAX_SIZE);
+      continue;
+    }
+
+    auto stream_name = context.getProperty(AmazonKinesisStreamName.name, flow_file.get());
+    if (!stream_name) {
+      logger_->log_error("Stream name is invalid due to {}", stream_name.error().message());
+      flow_file->setAttribute(AwsKinesisErrorMessage.name, fmt::format("Stream name is invalid due to {}", stream_name.error().message()));
+      session.transfer(flow_file, Failure);
+      continue;
+    }
+
+    auto entry = createEntryFromFlowFile(context, session, flow_file);
+    if (!entry) {
+      flow_file->addAttribute(AwsKinesisErrorMessage.name, entry.error().error_message);
+      if (entry.error().error_code) {
+        flow_file->addAttribute(AwsKinesisErrorCode.name, *entry.error().error_code);
+      }
+      session.transfer(flow_file, Failure);
+      continue;
+    }
+
+    auto [stream_batch, newly_created] = stream_batches.emplace(*stream_name, StreamBatch{});
+    if (newly_created) {
+      stream_batch->second.request.SetStreamName(*stream_name);
+    }
+    stream_batch->second.request.AddRecords(*entry);
+    stream_batch->second.items.push_back(BatchItem{.flow_file = std::move(flow_file), .error = std::nullopt});
+    stream_batches[*stream_name].batch_size += flow_file_size;
+    ++ff_count_in_batches;
+
+    if (stream_batches[*stream_name].batch_size > batch_data_size_soft_cap_) {
+      break;
+    }
+  }
+  return stream_batches;
+}
+
+void PutKinesisStream::processBatch(core::ProcessSession& session, StreamBatch& stream_batch, const Aws::Kinesis::KinesisClient& client) const {
+  const auto put_record_result = client.PutRecords(stream_batch.request);
+
+  auto transfer_failed_flow_files = gsl::finally([&] {
+    for (auto& [flow_file, error] : stream_batch.items) {
+      if (error) {
+        flow_file->addAttribute(AwsKinesisErrorMessage.name, error->error_message);
+        if (error->error_code) {
+          flow_file->addAttribute(AwsKinesisErrorCode.name, *(error->error_code));
+        }
+        session.transfer(flow_file, Failure);
+      }
+    }
+  });
```

Comment:
OK, this works, but the structure of the code is a bit hard to follow.  I would prefer to break the `processBatch()` function into two separate functions: one that calls `PutRecords()` and sets the errors, if any, and another one that iterates over the `stream_batch.items` and sets the flow file properties and does the transfers.